### PR TITLE
FSF-5965 | Velocity Planning - Working implementation, tests and launchfiles

### PR DESCRIPTION
**Pull request by DiogoProPrayer**
All the code was tested using the unit tests and was fixed according to the results of the tests. 
All the parameters were added to the launchfile.

### CODE REVIEW

File: src/planning/src/planning/velocity_planning.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  double y2 = point2.position.y;
+  double x3 = point3.position.x;
+  double y3 = point3.position.y;
+
+  PathPoint mid1 = PathPoint((x1 + x2) / 2, (y1 + y2) / 2, 0);
+  PathPoint mid2 = PathPoint((x2 + x3) / 2, (y2 + y3) / 2, 0);
+  double slope1 = (x2 != x1) ? ((y2 - y1) / (x2 - x1)) : MAXFLOAT;
+  double slope2 = (x3 != x2) ? ((y3 - y2) / (x3 - x2)) : MAXFLOAT;
+  double slope1_perpendicular = -1 / slope1;
+  double slope2_perpendicular = -1 / slope2;
```
Comment (lourenco31) [inline src/planning/src/planning/velocity_planning.cpp:17]: This breaks when y1=y2 or y3=y2. Please change to something like "slope1_perpendicular = (y1!=y2) ? -1.0 / slope1 : MINFLOAT; and the same thing for slope2_perpendicular

File: src/planning/include/planning/velocity_planning.hpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  
+    /**
+     * @brief function to calculate the radius of the circle that passes through 3 points
+     *
+     * @param point1 first point
+     * @param point2 second point
+     * @param point3 third point
+     * @return radius of the circle
+     */
+  double find_circle_center(PathPoint &point1, PathPoint &point2, PathPoint &point3);
```
Comment (lourenco31) [inline src/planning/include/planning/velocity_planning.hpp:42]: Shouldn't this be private?

File: src/planning/include/planning/velocity_planning.hpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * @brief function to limit the speed of the car according to the curvature of the lookahead path
+     *
+     * @param points path points
+     * @param velocities velocities of the path points
+     * @param brake_acelleration maximum braking acelleration
+     */
+  void speed_limiter(std::vector<PathPoint> &points, std::vector<double> &velocities);
```
Comment (lourenco31) [inline src/planning/include/planning/velocity_planning.hpp:51]: Shouldn't this also be private?

File: src/planning/include/planning/velocity_planning.hpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * @brief function to calculate the speed of the car according to the curvature of the path
+     * 
+     * @param radiuses radiuses vector of the path points
+     * @param velocities velocities vector of the path points
+     */
+    void point_speed(std::vector<double> &radiuses, std::vector<double> &velocities);
```
Comment (lourenco31) [inline src/planning/include/planning/velocity_planning.hpp:59]: Same thing

File: src/planning/src/planning/velocity_planning.cpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                     slope2_perpendicular * mid2.position.x + mid2.position.y - mid1.position.y) /
+                    (slope1_perpendicular - slope2_perpendicular);
+  double center_y = slope1_perpendicular * (center_x - mid1.position.x) + mid1.position.y;
+  double radius = sqrt(pow(center_x - x2, 2) + pow(center_y - y2, 2));
+  return radius;
+}
+
+void VelocityPlanning::point_speed(std::vector<double> &radiuses, std::vector<double> &velocities) {
+  for (int i = 0; i < static_cast<int>(radiuses.size()); i++) {
+    double velocity = sqrt(abs(config_.braking_acceleration_ * radiuses[i]));
```
Comment (lourenco31) [inline src/planning/src/planning/velocity_planning.cpp:28]: Maybe this was meant to be "normal_acceleration" instead of "braking_acceleration"

File: src/planning/src/planning/velocity_planning.cpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    velocities.push_back(std::max(velocity, config_.safety_speed_));
```
Comment (lourenco31) [inline src/planning/src/planning/velocity_planning.cpp:29]: Doesn't make much sense for me that if the velocity we calculate with sqrt(acceleration*radius) is really low, we go above it... is this right? It could be wiser to define a maximum speed and always go below it

File: src/planning/src/planning/velocity_planning.cpp [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  return;
+}
+
+void VelocityPlanning::speed_limiter(std::vector<PathPoint> &points,
+                                     std::vector<double> &velocities) {
+  for (int i = static_cast<int>(points.size()) - 2; i >= 0; i--) {
+    double distance = 0;
+    double max_speed = velocities[i];
+
+    for (int j = i + 1; j < static_cast<int>(points.size())-1; j++) {
```
Comment (lourenco31) [inline src/planning/src/planning/velocity_planning.cpp:42]: Are we skipping the last point in the path?

File: src/planning/src/planning/velocity_planning.cpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+      // Calculate segment distance
```
Comment (lourenco31) [inline src/planning/src/planning/velocity_planning.cpp:43]: I don't think we need to do this inner loop. Imagine we have an ordered path going from points P[0] to P[n]. 
We start by checking in the distance between P[n-1] and P[n] we can go from the maximum speed at P[n-1] to the maximum speed at P[n]. After that, when we check if we can get from the maximum speed at P[n-2] in the space between P[n-2] and P[n-1], we don't need to check the same thing for the speed at P[n] because we already know that from P[n-1] to P[n] there's no conflict between the maximum speeds. 

In short, I think we can do this in O(n)

File: src/planning/include/config/planning_config.hpp [hunk 9]
```diff
@@ -68,6 +74,11 @@ struct PlanningConfig {
 
     simulation_.publishing_visualization_msgs_ = params.publishing_visualization_msgs_;
     simulation_.using_simulated_se_ = params.using_simulated_se_;
+
+    velocity_planning_.safety_speed_ = params.minimum_velocity_;
```
Comment (lourenco31) [inline src/planning/include/config/planning_config.hpp:78]: It's not ideal to have two names for variables with the same purpose, I'd just chose one and go with it

File: src/planning/src/planning/velocity_planning.cpp (hunk 1 above)
Comment (DiogoProPrayer) [reply src/planning/src/planning/velocity_planning.cpp:17]: fixed with:
if ((slope1<=0.001) || (slope2<=0.001)) {
    return 10000;
  }

if the slope is 0, or very low, it will just return a large value because there is no need to spend resources calculating the real value

Comment (DiogoProPrayer) [reply src/planning/src/planning/velocity_planning.cpp:17]: I think this works:

double slope1 = (x2 != x1) ? (std::max((y2 - y1) / (x2 - x1), 0.0001)) : MAXFLOAT;
double slope2 = (x3 != x2) ? (std::max((y3 - y2) / (x3 - x2), 0.0001)) : MAXFLOAT;

File: src/planning/src/planning/velocity_planning.cpp [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+  double y1 = point1.position.y;
+  double x2 = point2.position.x;
+  double y2 = point2.position.y;
+  double x3 = point3.position.x;
+  double y3 = point3.position.y;
+
+  PathPoint mid1 = PathPoint((x1 + x2) / 2, (y1 + y2) / 2, 0);
+  PathPoint mid2 = PathPoint((x2 + x3) / 2, (y2 + y3) / 2, 0);
+  double slope1 = (x2 != x1) ? ((y2 - y1) / (x2 - x1)) : MAXFLOAT;
+  double slope2 = (x3 != x2) ? ((y3 - y2) / (x3 - x2)) : MAXFLOAT;
```
[the remark this answers is not in the document]
Comment (DiogoProPrayer) [reply src/planning/src/planning/velocity_planning.cpp:15]: Done

File: src/planning/include/planning/velocity_planning.hpp (hunk 3 above)
Comment (DiogoProPrayer) [reply src/planning/include/planning/velocity_planning.hpp:51]: ✅

File: src/planning/include/planning/velocity_planning.hpp (hunk 4 above)
Comment (DiogoProPrayer) [reply src/planning/include/planning/velocity_planning.hpp:59]: ✅

File: src/planning/src/planning/velocity_planning.cpp (hunk 7 above)
Comment (DiogoProPrayer) [reply src/planning/src/planning/velocity_planning.cpp:42]: Yes, the last point's velocity is fixed to the minimum velocity, it's the starting point to calculate the velocity for the other points

File: src/planning/src/planning/velocity_planning.cpp (hunk 6 above)
Comment (DiogoProPrayer) [reply src/planning/src/planning/velocity_planning.cpp:29]: If for some reason velocity planning breaks, the car will, at least, run at a minimal safety speed that we defined in the launch file. The safety speed should be lower than any speed caclulated by the program so it should never be the maximum, but if some reason it breaks at FSG, we can be sure that it will at least move slowly instead of not moving at all.

File: src/planning/include/config/velocity_config.hpp [hunk 11]
```diff
@@ -0,0 +1,37 @@
+#ifndef SRC_PLANNING_INCLUDE_CONFIG_VELOCITY_CONFIG_HPP_
+#define SRC_PLANNING_INCLUDE_CONFIG_VELOCITY_CONFIG_HPP_
+
```
[the remark this answers is not in the document]
Comment (DiogoProPrayer) [reply src/planning/include/config/velocity_config.hpp:3]: The "percentage" is defined in the launch file using the parameters normal_acceleration and braking_acceleration. If the car can withstand a normal acceleration of 15 m/s2 we can set how much we want to push it to its limit (for example: 66% of normal acceleration = 10m/s2). Using this aproach we can have a deeper understanding in where we are limiting the car (in cornering speed, or braking) instead of just a percentage. What do you think?

File: src/planning/src/planning/velocity_planning.cpp (hunk 8 above)
Comment (DiogoProPrayer) [reply src/planning/src/planning/velocity_planning.cpp:46]: Done

File: src/planning/include/config/velocity_config.hpp [hunk 12]
```diff
@@ -0,0 +1,37 @@
+#ifndef SRC_PLANNING_INCLUDE_CONFIG_VELOCITY_CONFIG_HPP_
+#define SRC_PLANNING_INCLUDE_CONFIG_VELOCITY_CONFIG_HPP_
```
Comment (Davide64-dev) [inline src/planning/include/config/velocity_config.hpp:2]: Please replace this with "pragma once"

File: src/planning/include/planning/velocity_planning.hpp [hunk 13]
```diff
@@ -0,0 +1,65 @@
+#ifndef SRC_PLANNING_INCLUDE_PLANNING_VELOCITY_HPP_
+#define SRC_PLANNING_INCLUDE_PLANNING_VELOCITY_HPP_
```
Comment (Davide64-dev) [inline src/planning/include/planning/velocity_planning.hpp:2]: idem

**Review by Davide64-dev — APPROVED** [review]

File: src/planning/src/planning/velocity_planning.cpp (hunk 1 above)
Comment (lourenco31) [reply src/planning/src/planning/velocity_planning.cpp]: That would make all negative slopes be calculated as 0.0001

**Review by lourenco31 — APPROVED** [review]
